File: src/scanner.cpp
#include "scanner.h"
#include "token_type.h"
#include "jaql.h"
#include <iostream>
#include <any>
#include <cctype>

namespace Jaql
{
    Scanner::Scanner(std::string source):
    source(std::move(source))
    {
        reserved_words = {
            {"and", TokenType::AND},
            {"class", TokenType::CLASS},
            {"else", TokenType::ELSE},
            {"false", TokenType::FALSE},
            {"for", TokenType::FOR},
            {"fun", TokenType::FUN},
            {"if", TokenType::IF},
            {"nil", TokenType::NIL},
            {"or", TokenType::OR},
            {"print", TokenType::PRINT},
            {"return", TokenType::RETURN},
            {"super", TokenType::SUPER},
            {"this", TokenType::THIS},
            {"true", TokenType::TRUE},
            {"var", TokenType::VAR},
            {"while", TokenType::WHILE},
        };
    }


    std::vector<Token> Scanner::scan_tokens()
    {
        while (!is_at_end()){
            start = current;
            scan_token();
        }
        tokens.emplace_back(TokenType::EOJF, "", line);

        for(const auto& token: tokens) {
            std::cout << token.to_string() << "\n";
        }
        return tokens;
    }

    bool Scanner::is_at_end()
    {
        return current >= static_cast<int>(source.size());
    }
    
    void Scanner::scan_token()
    {
        char c = advance();
        switch (c){
            case '(': add_token(TokenType::LEFT_PAREN); break;
            case ')': add_token(TokenType::RIGHT_PAREN); break;
            case '{': add_token(TokenType::LEFT_BRACE); break;
            case '}': add_token(TokenType::RIGHT_BRACE); break;
            case ',': add_token(TokenType::COMMA); break;
            case '.': add_token(TokenType::DOT); break;
            case '-': add_token(TokenType::MINUS); break;
            case '+': add_token(TokenType::PLUS); break;
            case ';': add_token(TokenType::SEMICOLON); break;
            case '*': add_token(TokenType::STAR); break; 
            case '!':
                add_token(match('=') ? TokenType::BANG_EQUAL : TokenType::BANG);
                break;
            case '=':
                add_token(match('=') ? TokenType::EQUAL_EQUAL : TokenType::EQUAL);
                break;
            case '<':
                add_token(match('=') ? TokenType::LESS_EQUAL : TokenType::LESS);
                break;
            case '>':
                add_token(match('=') ? TokenType::GREATER_EQUAL : TokenType::GREATER);
                break;
            case '/':
                if (match('/')){
                    while (peek() != '\n' && !is_at_end()) advance();
                } else{
                    add_token(TokenType::SLASH);
                }

            case ' ':
            case '\r':
            case '\t':
                // Ignore whitespace.
                break;

            case '\n':
                ++line;
                break;
            
            case '"': string(); break;

            default:
                if (std::isdigit(c))
                {
                    number();
                } else if (std::isalpha(c) || c == '_'){
                    identifier();
                }
                else{
                    Jaql::error(line, "Unexpected character.");
                    break;
                }
        }
    }

    void Scanner::identifier()
    {
        while ( std::isdigit(peek()) || std::isalpha(peek()) || peek() == '_' )
        {
            advance();
        }
        std::string ident = source.substr(start, current-start);

        if (const auto id=reserved_words.find(ident); id != reserved_words.end()){
            add_token(id->second);
        }else{
            add_token(TokenType::IDENTIFIER);
        }
        
    }

    void Scanner::number()
    {
        while (std::isdigit(peek())) advance();

        if (peek() == '.' && std::isdigit(peek_next())){
            advance();

            while (std::isdigit(peek())) advance();
        }
        add_token(TokenType::NUMBER, std::stod(source.substr(start, current-start)) );
    }

    char Scanner::peek_next(){
        if (current + 1 >= source.length()) return '\0';
        return source.at(current+1);
    }

    void Scanner::string()
    {
        while (peek() != '"' && !is_at_end()) {
            if (peek() == '\n') ++line;
            advance();
        }

        if (is_at_end()) {
            Jaql::error(line, "Unterminated string.");
            return;
        }

        // The closing ".
        advance();

        // Trim the surrounding quotes.
        std::string value = source.substr(start + 1, current - 2);
        add_token(TokenType::STRING, value);
    }

    char Scanner::peek()
    {
        if (is_at_end()) return '\0';
        return source.at(current);
    }

    bool Scanner::match(char expected)
    {
        if (is_at_end()) return false;
        if (source.at(current) != expected ) return false;
        ++current;
        return true;
    }

    char Scanner::advance()
    {
        ++current;
        return source.at(current-1);
    }

    void Scanner::add_token(TokenType type)
    {
        add_token(type, NULL);
    }

    void Scanner::add_token(TokenType type, std::any literal)
    {
        tokens.emplace_back(type, source.substr(start, current-start), literal, line);
    }
};

File: src/scanner.py
import logging

from src.exceptions import JaqlException
from src.jaql import Jaql
from src.token import Token
from src.token_type import TokenType


class Scanner:
    def __init__(self, source: str, jaql: Jaql, debug=False) -> None:
        self.source = source
        self.jaql = jaql
        self.debug = debug
        self.size = len(source)
        self.current = 0
        self.start = 0
        self.line = 1
        self.tokens: list[Token] = []
        self.reserved_words = {
            "and": TokenType.AND,
            "class": TokenType.CLASS,
            "else": TokenType.ELSE,
            "false": TokenType.FALSE,
            "for": TokenType.FOR,
            "fun": TokenType.FUN,
            "if": TokenType.IF,
            "nil": TokenType.NIL,
            "or": TokenType.OR,
            "print": TokenType.PRINT,
            "return": TokenType.RETURN,
            "super": TokenType.SUPER,
            "this": TokenType.THIS,
            "true": TokenType.TRUE,
            "var": TokenType.VAR,
            "while": TokenType.WHILE,
        }

    def is_at_end(self):
        return self.current >= self.size

    def advance(self):
        self.current += 1
        return self.source[self.current - 1]

    def add_token(self, token_type: TokenType, literal=None):
        self.tokens.append(
            Token(
                token_type,
                self.source[self.start : self.current],
                self.line,
                literal,
            )
        )

    def peek(self):
        if self.is_at_end():
            return "\0"
        return self.source[self.current]

    def peek_next(self):
        if (self.current + 1) >= self.size:
            return "\0"
        return self.source[self.current + 1]

    def match(self, expected):
        if self.is_at_end():
            return False
        if self.source[self.current] != expected:
            return False
        self.current += 1
        return True

    def string(self):
        while self.peek() != '"' and not self.is_at_end():
            if self.peek() == "\n":
                self.line += 1
            self.advance()

        if self.is_at_end():
            # TODO: raise exceptions up the stack proper
            # Maybe want the main jaql to inherit everything?
            self.jaql.add_error(self.line, "Unterminated sring")

        self.advance()
        value = self.source[self.start + 1 : self.current - 1]
        self.add_token(TokenType.STRING, literal=value)

    def number(self):
        while self.peek().isdigit():
            self.advance()
            if self.peek() == "." and self.peek_next().isdigit():
                self.advance()
        self.add_token(TokenType.NUMBER, float(self.source[self.start : self.current]))

    def identifier(self):
        while self.peek().isalnum() or self.peek() == "_":
            self.advance()
        ident = self.source[self.start : self.current]

        if token_type := self.reserved_words.get(ident):
            self.add_token(token_type)
        else:
            self.add_token(TokenType.IDENTIFIER)

    def scan_token(self):
        c = self.advance()
        match c:
            case "(":
                self.add_token(TokenType.LEFT_PAREN)
            case ")":
                self.add_token(TokenType.RIGHT_PAREN)
            case "{":
                self.add_token(TokenType.LEFT_BRACE)
            case "}":
                self.add_token(TokenType.RIGHT_BRACE)
            case ",":
                self.add_token(TokenType.COMMA)
            case ".":
                self.add_token(TokenType.DOT)
            case "-":
                self.add_token(TokenType.MINUS)
            case "+":
                self.add_token(TokenType.PLUS)
            case ";":
                self.add_token(TokenType.SEMICOLON)
            case "*":
                self.add_token(TokenType.STAR)
            case "!":
                if self.match("="):
                    token_type = TokenType.BANG_EQUAL
                else:
                    token_type = TokenType.BANG
                self.add_token(token_type)

            case "=":
                if self.match("="):
                    token_type = TokenType.EQUAL_EQUAL
                else:
                    token_type = TokenType.EQUAL
                self.add_token(token_type)

            case "<":
                if self.match("="):
                    token_type = TokenType.LESS_EQUAL
                else:
                    token_type = TokenType.LESS
                self.add_token(token_type)

            case ">":
                if self.match("="):
                    token_type = TokenType.GREATER_EQUAL
                else:
                    token_type = TokenType.GREATER
                self.add_token(token_type)

            case "/":
                if self.match("/"):
                    while self.peek() != "\n" and not self.is_at_end():
                        self.advance()
                else:
                    self.add_token(TokenType.SLASH)

            case " ":
                pass
            case "\r":
                pass
            case "\t":
                # Ignore whitespace.
                pass

            case "\n":
                self.line += 1

            case '"':
                self.string()

            case _:
                if c.isdigit():
                    self.number()
                elif c.isalpha() or c == "_":
                    self.identifier()
                else:
                    self.jaql.add_error(self.line, "Unexpected character.")

    def scan_tokens(self):
        while not self.is_at_end():
            self.start = self.current
            self.scan_token()

        self.tokens.append(Token(TokenType.EOJF, "", self.line))

        if self.debug:
            for token in self.tokens:
                print(token.to_string())
        return self.tokens
